vncviewer/GestureHandler.cxx

```diff
@@ -0,0 +1,364 @@
+/* Copyright 2019 Aaron Sowry for Cendio AB
+ * 
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#include <cmath>
+
+#include <rfb/LogWriter.h>
+
+#include <X11/extensions/XI2.h>
+#include <X11/extensions/XInput2.h>
+
+#include "GestureHandler.h"
+
+static rfb::LogWriter vlog("GestureHandler");
+
+// Movement threshold for gestures
+const unsigned GH_MTHRESHOLD = 50;
+
+// Timeout when waiting for gestures (ms)
+const unsigned GH_STTDELAY = 250;
+
+GestureHandler::GestureHandler() :
+  state(GH_INITSTATE), timeoutTimer(this) {
+}
+
+GestureHandler::~GestureHandler()
+{
+}
+
+void GestureHandler::registerEvent(const XIDeviceEvent *devev) {
+
+  switch (devev->evtype) {
+    case XI_TouchBegin:
+      // Ignore any new touches if there is already an active gesture
+      if (!hasDetectedGesture())
+        trackTouch(devev);
```

Comment:
I'm still a bit confused here as to what you mean by "unexpected" touches. Do you mean "residual" touches perhaps? For example, a vertical scroll gesture ends as soon as one finger is removed from the screen. That means there could still be one finger left on the screen after the scroll has ended and we're ready to start accepting new gestures. Do you mean that we should not start accepting a new gesture until all fingers have been removed?

Reply:
More or less. Perhaps the user unintentionally lifted that finger, or there was some glitch in the device's detection. There could also be other mistakes that we ignored (because a gesture was already detected) that could cause confusion. E.g. the user accidentally overlapped two gestures, partially or fully.

In all such weird cases I think it would be better if we can end up doing nothing, rather than doing something random (from the user's point of view).

Reply:
I think this could be done without having to track ignored touches separately. What if we just removed touches from `tracked` one at a time as the fingers are removed, and don't start accepting new touches until it's empty?

I've done some testing, and the only weird behaviour I can reproduce is the situation I mentioned above, with residual fingers remaining from a completed gesture. I'm not convinced this is a big enough issue to warrant the added complexity (and time..) of fixing it. Another enhancement for later?

Reply:
From what I understand, this looks half-solved in the current code. We place ignored touches in a list, but never use that list for anything? How is this supposed to work?

Reply:
It's checked on touch releases to see if we can reset detection.

Reply:
Ah, I see now, thanks.